Fix logging from wptserve when using wptrunner

jgraham opened the pull request.

jgraham committed:
Reduce the amount of error-level logging wptserve emits for tracebacks

This is enough to indicate there's an error, but not so much we swamp log
summarisation

jgraham committed:
Use stdlib logger everywhere in wptserve and serve

Previously the integration between wptrunner and wptserve involved
tyring to use a mozlog logger in wptserve when initiated from
wptrunner. This was complex and prone to failure (it has been somewhat
broken for a long time).

To simplify adopt the following approach:

* wptserve and serve no longer have configurable loggers. It's still
possible to set the log level, but the logger itself is removed from
the configuration and the stdlib logging module root logger is always
used instead (we could consider using a named logger rather than the
root, but this doesn't make much difference).

* When starting servers in multiple processses, the logging is always
configured at the start of each process. We don't explicitly store the
logger as a global anywhere.

* For integration with mozlog/wptrunner, we create a handler that
converts stdlib LogRecord object into mozlog format and puts it on a
queue. Then the main process has a thread which drains this queue to
the logger. This handler needs to be passed down to all the points
where logging can be configured (notably the ServerProc.create_daemon
function, which is where we start the processes running servers).

jgraham committed:
Don't try to handle exceptions directly in cookies script

jgraham committed:
Make cookies testdriver test encode as JSON

This is required by the server-side script

gsnedders commented on tools/wptserve/wptserve/config.py:
```diff
@@ -61,12 +70,17 @@ def __len__(self):
 
     def as_dict(self):
-        return json_types(self.__dict__)
+        data = self.__dict__.copy()
```
quite probably we want a deep copy here, no?

gsnedders commented on tools/wptserve/wptserve/config.py:
```diff
@@ -28,11 +28,20 @@ def _merge_dict(base_dict, override_dict):
+    :param data: - Extra configuration data
+    """
+    def __init__(self, logger, data):
+        if isinstance(logger, str):
```
It also feels a bit surprising that logger is a str/logger? Can we make the ConfigBuilder always return a (safe) logger object?

foolip replied:
Is the contract something like the parent process picks a unique logger name, hands that to wptserve, and then parses the stdout to forward the stuff logged with that logger name? And there's the assumption that wptserve is using mozlog with a specific format I guess?

gsnedders commented on tools/wptserve/wptserve/config.py:
```diff
@@ -28,11 +28,20 @@ def _merge_dict(base_dict, override_dict):
+    Inherits from Mapping for backwards compatibility with the old dict-based config
+
+    :param logger: - Either the name of the logger to use, or an object that will be
+                     used as the logger. In the case of an oject this must be safe
```
```suggestion
                     used as the logger. In the case of an object this must be safe
```

gsnedders commented on tools/wptrunner/wptrunner/environment.py:
```diff
@@ -50,6 +51,74 @@ class TestEnvironmentError(Exception):
+class LoggerProxy:
+    """Logger frontend that puts messages on a provided queue.
+
+    This is intended for use in wptserve which doesn't directly depend
+    on mozlog, so it presents an API like the stdlib logger module.
```
Maybe something more like:

> This provides an API-compatible implementation of the stdlib logging.Logger. It is intended for use in dependencies using the logging module, allowing logs to be forwarded to mozlog as needed.

gsnedders commented on tools/wptrunner/wptrunner/environment.py:
```diff
@@ -50,6 +51,74 @@ class TestEnvironmentError(Exception):
+
+    This is intended for use in wptserve which doesn't directly depend
+    on mozlog, so it presents an API like the stdlib logger module.
+    However to allow the data to be used directly by mozlog, we construct
```
New paragraph, maybe drop the "However"?

gsnedders commented on tools/wptrunner/wptrunner/environment.py:
```diff
@@ -50,6 +51,74 @@ class TestEnvironmentError(Exception):
+    """Context manager object that handles setup and teardown of a logger to pass to wptserve"""
+
+    def __init__(self):
+        server_logger = get_default_logger(component="wptserve")
```
Can we avoid hard-coding this and make it a name argument?

gsnedders commented on tools/wptrunner/wptrunner/environment.py:
```diff
@@ -50,6 +51,74 @@ class TestEnvironmentError(Exception):
+        self._logger.debug(*args)
+
+
+class ServerLogging:
```
Aside from the hard-coded strings, there's nothing really server-specific about this. Consider renaming.

jgraham commented on tools/wptserve/wptserve/response.py:
```diff
@@ -243,7 +243,13 @@ def set_error(self, code, message=u""):
+                first_line = message.splitlines()[0]
+            else:
+                first_line = "<no message given>"
+            self.logger.error("Exception loading %s: %s" % (self.request.url,
```
I think it can be used for non-500 errors too (it's part of the response API so handlers can call it) and in practice there's a single exception handler in the request handling function that calls this. In any case out of scope for this patch.

foolip commented on tools/wptrunner/wptrunner/environment.py:
```diff
@@ -50,6 +51,76 @@ class TestEnvironmentError(Exception):
+        super().__init__(name)
+        assert not self.handlers
+
+    def _handle_log(self, data):
```
What ends up calling `_handle_log`? Something in mozlog?

jgraham replied:
Yes, https://searchfox.org/mozilla-central/source/testing/mozbase/mozlog/mozlog/structuredlog.py#278

foolip commented on tools/wptrunner/wptrunner/environment.py:
```diff
@@ -50,6 +51,76 @@ class TestEnvironmentError(Exception):
+        name = str(uuid.uuid4())
+        self.queue = queue
+        super().__init__(name)
+        assert not self.handlers
```
What's the significance of this assert? Is it that `super().__init__(name)` shouldn't initialize `self.handlers`? Something does by the time it's used, but does it matter that it wasn't here?

foolip commented on tools/wptrunner/wptrunner/environment.py:
```diff
@@ -50,6 +51,76 @@ class TestEnvironmentError(Exception):
 
+class _QueueLogger(StructuredLogger):
+    """Structured logger that has a unique name and puts messages on a queue"""
+    def __init__(self, queue):
```
```suggestion
    def __init__(self, queue: QueueOfSomeKind):
```

I'm having trouble finding what `queue` is, and this new code seems like a fine place to start using type annotations. I don't know what the actual type is here, but I'll suggest the types where I think it's not obvious.

foolip commented on tools/wptrunner/wptrunner/environment.py:
```diff
@@ -50,6 +51,76 @@ class TestEnvironmentError(Exception):
+class ProxyLoggingContext:
+    """Context manager object that handles setup and teardown of a logger to pass to wptserve"""
+
+    def __init__(self, component):
```
```suggestion
    def __init__(self, component: str):
```

foolip commented on tools/wptserve/wptserve/config.py:
```diff
@@ -28,11 +28,25 @@ def _merge_dict(base_dict, override_dict):
-        self.__dict__["_logger_name"] = logger_name
+    Immutable configuration that's safe to be passed between processes.
+
+    Inherits from Mapping for backwards compatibility with the old dict-based config
```
In other words, to support code that uses `config["key"]`, right? Is there any new way too, or actually just the old way?

foolip replied:
So my question was if the "old dict-based config" is retained here because there are call sites to it, or if it just falls out of a reasonable way to implement this? The comment makes me wonder where the badness is. Not that all old things are bad, but in code comments it's usually disparaging :)

gsnedders replied:
IIRC, there's various handlers that read from it as if it were a dict. Ultimately, the cost of supporting both is very small.

foolip commented on tools/wptserve/wptserve/config.py:
```diff
@@ -28,11 +28,25 @@ def _merge_dict(base_dict, override_dict):
+                     to share directly between processes.
+    :param data: - Extra configuration data
+    """
+    def __init__(self, logger, data):
```
```suggestion
    def __init__(self, logger: Union[str, SomethingElse], data):
```

This would be very helpful to understand what the `SomethingElse` that gets set on `self.__dict__["_logger"]` is.

foolip replied:
Aha, so it's never a string, always an object that has certain methods like `logger.warn` and so on? That's what I couldn't really figure out. A comment saying that this needs to be a cross-process serializable thing would be great.

foolip commented on tools/wptserve/wptserve/config.py:
```diff
@@ -61,12 +75,14 @@ def __len__(self):
 
     @property
     def logger(self):
```
```suggestion
    def logger(self) -> Something
```

would be helpful here. Or just a comment that says what this thing is. In particular I'm wondering if it's definitely a JSON-serializable value, or what is the thing getting passed between processes?

jgraham replied:
No, this is "a thing that implements the logging API". It's specifically not required to be serializable; the logger is created in the current process (in the case when a name is supplied).

foolip replied:
OK, so a comment block in a central place explaining how this is all wired together would be helpful. My naive question when arriving at that comment block would be "why doesn't the subprocess just log to stdout and the parent process just copies it line-by-line?"

foolip commented on tools/wptserve/wptserve/config.py:
```diff
@@ -134,6 +152,25 @@ class ConfigBuilder(object):
+
+    :param logger: - A logger object. The name and level of this logger
+                     will be part of the configuration, but the object
+                     iself is not included.
```
```suggestion
                     itself is not included.
```

foolip commented on tools/wptserve/wptserve/config.py:
```diff
@@ -134,6 +152,25 @@ class ConfigBuilder(object):
+    :param logger: - A logger object. The name and level of this logger
+                     will be part of the configuration, but the object
+                     iself is not included.
+    :param subdomains: - A set of valid subdomains to included in the
```
```suggestion
    :param subdomains: - A set of valid subdomains to include in the
```

foolip commented on tools/wptserve/wptserve/config.py:
```diff
@@ -134,6 +152,25 @@ class ConfigBuilder(object):
+                     iself is not included.
+    :param subdomains: - A set of valid subdomains to included in the
+                         configuration.
+    :param not_subdomains: - A set of invalid subdomains to included in
```
```suggestion
    :param not_subdomains: - A set of invalid subdomains to include in
```

foolip commented on tools/wptserve/wptserve/config.py:
```diff
@@ -186,20 +223,28 @@ def __init__(self,
-            self._logger_name = logger.name
+            self._proxy_logger = None
+            if logger is None:
+                self._logger_name = "web-platform-tests"
```
Does this need to be unique? If not, maybe "wptserve" gives a better clue? Or if it doesn't appear anywhere at all, just the empty string :)

foolip reviewed (commented):
I don't entirely understand what's going on, but I can ask questions and point out typos :)

jgraham commented:
I changed the implementation a bit in response to feedback. idk to what extent we want to go all in on the types here; to some extent it seems like the annotations are going to be misleading until we support mypy for these parts, which is going to require more effort to add annotations.

jgraham committed:
Ensure that wptserve shuts down subprocesses cleanly

We were calling terminate() on the server processes, which meant that
they didn't have a chance to shutdown properly and in particular could
end up failing to unlock cross-process shared locks.

This replaces the kill() method with a stop() method that arranges for
a clean shutdown.

It also removed the block argument from the Daemon process start
argument, since running the servers in a way that directly blocks the
main loop isn't really supported. Instead we always arrange for the
server to be started in a thread so we're later able to call
shutdown() on it.

There was also some signal handling logic that wasn't really doing
much, so that's replaced with just trapping the default exceptions.

foolip commented:
I've looked at each individual commit and this looks good! I have some nits/questions, some on existing code, which I'll follow up with.

foolip commented on tools/wptrunner/wptrunner/wptlogging.py:
```diff
@@ -65,3 +66,41 @@ def __call__(self, data):
+
+    def createLock(self):
+        # The queue provides its own locking
+        self.lock = None
```
If it’s not set to None, will something else set it? Would an assert suffice?

jgraham replied:
See https://github.com/python/cpython/blob/main/Lib/logging/__init__.py#L2211 `createLock` actually has to create the lock, but the type is effectively `Nullable[Lock]` so this is correct.

gsnedders commented on infrastructure/testdriver/delete_all_cookies.html:
```diff
@@ -17,21 +17,11 @@
-  await fetch(`/cookies/resources/cookie.py?set=${encoded4}`)
+  const cookies = ["test1=1", "test2=2; path=/", "test3=3; path=/cookies/resources", "test4=4; HttpOnly"];
+  for (const cookie of cookies) {
+      const encoded = JSON.stringify(encodeURIComponent(cookie));
```
Do we not want `encodeURIComponent(JSON.stringify(`…?

gsnedders commented on tools/serve/serve.py:
```diff
@@ -500,21 +499,26 @@ def __init__(self, mp_context, scheme=None):
-                      config, **kwargs):
-        ensure_logger(config)
+                      config, log_handlers, **kwargs):
+        # Ensure that when we start this in a new process we have the global lock
```
And what global lock? One of ours, or belonging to the stdlib?

foolip replied:
OK, there were 3 instances of this comment and now just one. That's an improvement :)

After reading https://docs.python.org/3/library/importlib.html#importlib.reload I gather that whatever internal state the logging module had will be reset. Maybe append "Reloading the module will release the lock if it was held"?

gsnedders commented on tools/wptrunner/wptrunner/environment.py:
```diff
@@ -50,16 +49,47 @@ class TestEnvironmentError(Exception):
+    logger = get_default_logger(component="wptserve")
+    log_filter = handlers.LogLevelFilter(lambda x: x, "info")
+    # Downgrade errors to warnings for the server
+    log_filter = LogLevelRewriter(log_filter, ["error"], "warning")
```
Their omission here matches the existing behaviour, and I think is desired because we rarely raise critical errors and they tend to suggest something is very wrong.

foolip replied:
OK, if this was existing behavior (or at least *intended* behavior given that it's current broken) then let's leave that.

gsnedders commented on tools/serve/serve.py:
```diff
@@ -543,23 +543,19 @@ def create_daemon(self, init_func, host, port, paths, routes, bind_address,
+                self.daemon.start()
                 try:
-                    self.stop.wait()
+                    self.stop_flag.wait()
```
Do we want a timeout here?

jgraham replied:
I don't think that makes sense? The stop flag is an IPC mechanism to signal to the server that it should shut down; the process owning the server waits on the flag whilst running the server itself in another thread. I suppose the failure mode here is "the other thread stopped and we're still waiting", but it would be a larger change to fix that (we'd need to add a method to the server objects to check the serving thread is alive and poll that periodically, or something).

gsnedders commented on tools/serve/serve.py:
```diff
@@ -543,23 +543,19 @@ def create_daemon(self, init_func, host, port, paths, routes, bind_address,
                 try:
-                    self.stop.wait()
+                    self.stop_flag.wait()
+                    self.daemon.stop()
```
Should we have this in a `finally` block instead, in case the above throws (somehow)?

jgraham commented on tools/serve/serve.py:
```diff
@@ -1110,36 +1087,40 @@ def handle_signal(signum, frame):
-                for subproc in iter_procs(servers):
-                    subproc.join(1)
+            try:
+                while (all(server.proc.is_alive() for server in iter_servers(servers)) and
```
I think the idea is that if one server fails to start, or otherwise shuts down, we want to shut down everything rather than keeping going in a partially functional state.

jgraham replied:
I've rearranged the code in a way that's hopefully clearer.

foolip replied:
Aha, that's *much* clearer. I thought this was the cleanup code on shutdown, and it seemed very weird wait for servers exit if *all* of them were still running as opposed to if *any* of them were...

jgraham commented on tools/wptrunner/wptrunner/environment.py:
```diff
@@ -50,16 +49,47 @@ class TestEnvironmentError(Exception):
+
+    def __exit__(self, *args):
+        self.log_queue.put(None)
+        # Wait for thread to shut down but not for too long since it's a daemon
```
Well in theory "nothing"; it's a daemon thread so if we exit the program it won't block waiting for it to end. In practice, if it hasn't shut down in one second that either means:
a) There's such a backlog of logging messages that we haven't processed them inside a second and so didn't process the `None` message which triggers shutdown
b) There's some kind of deadlock affecting the queue object and we're going to hang at shutdown.

It's honestly somewhat unclear to me if it makes any meaningful difference if we specify a timeout here.

foolip replied:
If it's not clear if this is needed, and it's untested, then just waiting indefinitely for the logging thread to exit seems slightly better for discovering if something's actually wrong?

jgraham commented on tools/wptrunner/wptrunner/tests/test_products.py:
```diff
@@ -49,7 +49,7 @@ def test_server_start_config(product):
             start.assert_called_once()
             args = start.call_args
-            config = args[0][0]
+            config = args[0][1]
```
We changed the signature of `serve.start`so the logger is argument 0 and the config is argument 1.

foolip commented on tools/wptserve/wptserve/logger.py:
```diff
@@ -1,29 +1,5 @@
-class NoOpLogger(object):
```
Glad to see this go, it came up in mypy changes I was working on :)

foolip reviewed (approved):
Some small nits, but I'm happy for this to land without taking another look.

foolip commented on tools/serve/wave.py:
```diff
@@ -3,7 +3,6 @@
 import subprocess
 from manifest import manifest
 import localpaths
-import logging
```
I'm on macOS, so maybe some multiprocessing defaults are different?

jgraham replied:
https://github.com/web-platform-tests/wpt/pull/29051 although mac should default to spawn from 3.8 onwards.

foolip replied:
I'll try if that fixes the problem for me. I'm already on 3.8 but it's the system Python so it might not be identical to vanilla Python, who knows?
